src/regexset.rs

```diff
@@ -0,0 +1,832 @@
+            let flags = self.compute_flags();
+            let mut expr_tree = Expr::parse_tree_with_flags(pattern, flags)?;
+
+            // Try to optimize the expression tree
+            let requires_capture_group_fixup = optimize(&mut expr_tree);
+            let info = analyze(&expr_tree, requires_capture_group_fixup)?;
+
+            if info.hard {
+                // Hard pattern - compile to VM program
+                let prog = compile(&info, can_compile_as_anchored(&expr_tree.expr))?;
+                let pattern_options = RegexOptions {
+                    pattern: pattern.clone(),
+                    syntaxc: self.syntaxc,
+                    backtrack_limit: self.backtrack_limit,
+                    delegate_size_limit: self.delegate_size_limit,
+                    delegate_dfa_size_limit: self.delegate_dfa_size_limit,
+                    oniguruma_mode: false,
+                };
+                hard_patterns.push(HardPattern {
+                    pattern_id: index,
+                    prog: Arc::new(prog),
+                    n_groups: info.end_group(),
+                    named_groups: Arc::new(expr_tree.named_groups),
+                    options: pattern_options,
+                });
+            } else {
+                // Easy pattern - extract delegate string for DFA
+                let mut delegate_str = String::new();
+                expr_tree.expr.to_str(&mut delegate_str, 0);
+                easy_pattern_strings.push(delegate_str.clone());
+                easy_pattern_infos.push(EasyPatternInfo {
+                    pattern_id: index,
+                    delegate_pattern: delegate_str,
+                    named_groups: Arc::new(expr_tree.named_groups),
+                    explicit_capture_group_0: requires_capture_group_fixup,
+                });
+            }
+        }
+
+        // Build multi-pattern DFA for easy patterns
+        let easy_patterns = if !easy_pattern_strings.is_empty() {
+            let mut config = RaConfig::new();
+            if let Some(limit) = self.delegate_size_limit {
+                config = config.nfa_size_limit(Some(limit));
+            }
+            if let Some(limit) = self.delegate_dfa_size_limit {
+                config = config.dfa_size_limit(Some(limit));
+            }
+
+            let dfa = RaBuilder::new()
+                .configure(config)
+                .syntax(self.syntaxc)
+                .build_many(&easy_pattern_strings)
+                .map_err(crate::CompileError::InnerError)?;
+
+            Some(EasyPatternSet {
+                dfa,
+                patterns: easy_pattern_infos,
+            })
+        } else {
+            None
+        };
+
+        Ok(RegexSet {
+            inner: Arc::new(RegexSetImpl {
+                easy_patterns,
+                hard_patterns,
+                backtrack_limit: self.backtrack_limit,
+            }),
+        })
+    }
+
+    fn compute_flags(&self) -> u32 {
```

Comment:
Maybe we can somehow have both builders share the functions for setting flags and computing them?